go/apps/api/routes/v2_apis_list_keys/handler.go

```diff
@@ -158,319 +154,208 @@ func (h *Handler) Handle(ctx context.Context, s *zen.Session) error {
-		for _, key := range keys {
-			if !key.IdentityID.Valid {
-				continue
-			}
-
-			uniqIdentityIds[key.IdentityID.String] = struct{}{}
-		}
-
-		identityIDs := make([]sql.NullString, 0)
-		for identityID := range uniqIdentityIds {
-			identityIDs = append(identityIDs, sql.NullString{String: identityID, Valid: true})
-		}
-
-		// Query ratelimits for these identities
-		identityRatelimits, listErr := db.Query.ListIdentityRatelimitsByIDs(ctx, h.DB.RO(), identityIDs)
-		if listErr != nil {
-			return fault.Wrap(listErr,
-				fault.Code(codes.App.Internal.ServiceUnavailable.URN()),
-				fault.Internal("database error"), fault.Public("Failed to retrieve ratelimits."),
-			)
-		}
-
-		// Group ratelimits by identity_id
-		for _, rl := range identityRatelimits {
-			if rl.IdentityID.Valid {
-				identityRatelimitsMap[rl.IdentityID.String] = append(identityRatelimitsMap[rl.IdentityID.String], rl)
-			}
-		}
-	}
-
-	// If user requested decryption, check permissions and decrypt
-	plaintextMap := map[string]string{} // nolint:staticcheck
+	// Handle decryption if requested
+	plaintextMap := make(map[string]string)
 	if req.Decrypt != nil && *req.Decrypt {
-		// If we have permission, proceed with decryption
-		for _, key := range keys {
+		for _, key := range keyResults {
 			if key.EncryptedKey.Valid && key.EncryptionKeyID.Valid {
 				decrypted, decryptErr := h.Vault.Decrypt(ctx, &vaultv1.DecryptRequest{
-					Keyring:   key.Key.WorkspaceID,
+					Keyring:   key.WorkspaceID,
 					Encrypted: key.EncryptedKey.String,
 				})
 				if decryptErr != nil {
 					h.Logger.Error("failed to decrypt key",
-						"keyId", key.Key.ID,
+						"keyId", key.ID,
 						"error", decryptErr,
 					)
 					continue
 				}
-
-				plaintextMap[key.Key.ID] = decrypted.GetPlaintext()
+				plaintextMap[key.ID] = decrypted.GetPlaintext()
 			}
 		}
 	}
 
-	// Determine the cursor for the next page
-	hasMore := len(keys) > limit
-	var nextCursor *string
-	if hasMore {
-		nextCursor = ptr.P(keys[len(keys)-1].Key.ID)
-		// Trim the results to the requested limit
-		keys = keys[:limit]
-	}
-
-	// Transform keys into the response format
-	responseData := make([]openapi.KeyResponseData, len(keys))
-	for i, key := range keys {
-		k := openapi.KeyResponseData{
-			KeyId:       key.Key.ID,
-			Start:       key.Key.Start,
-			CreatedAt:   key.Key.CreatedAtM,
-			Enabled:     key.Key.Enabled,
-			Credits:     nil,
-			Expires:     nil,
-			Identity:    nil,
-			Meta:        nil,
-			Name:        nil,
-			Plaintext:   nil,
-			UpdatedAt:   nil,
-			Ratelimits:  nil,
-			Permissions: nil,
-			Roles:       nil,
-		}
-
-		if key.Key.Name.Valid {
-			k.Name = ptr.P(key.Key.Name.String)
+	// Transform to response format
+	responseData := make([]openapi.KeyResponseData, len(keyResults))
+	for i, key := range keyResults {
+		keyRow := db.FindLiveKeyByIDRow(key)
+		keyData := db.ToKeyData(keyRow)
+		response, err := h.BuildKeyResponseData(keyData, plaintextMap[key.ID])
```

Comment:
I'm not sure we need to extract this into a seperate function we can just do it here and return the error directly. 

even tho it is unlikely those errors case will ever be hit. I am not even sure we should fail when a key has invalid json meta because this would block the user from listing any keys where this key is in range.